scripts/linters/pre_commit_linter.py

```diff
@@ -471,23 +483,22 @@ def main(args=None):
     tasks_third_party = []
 
     for linter in custom_linters:
-        # Concurrency limit: 25.
         task_custom = concurrent_task_utils.create_task(
-            linter.perform_all_lint_checks, 25,
-            verbose=verbose_mode_enabled, name='custom')
+            linter.perform_all_lint_checks, verbose_mode_enabled,
+            custom_semaphore, name='custom')
```

Comment:
I understand why you need the semaphore in execute_tasks, but why does create_task need it?

In general I think you should find a way to avoid repeating the need to pass the semaphore in two different places. There's a hidden constraint implied there (the semaphore arg must be the same for create_task and execute_tasks) which is not enforced anywhere. This makes the code hard to maintain.

Reply:
Actually, you'll need to pass it into both. The rationale is this:

- The executor needs a reference so that it can try to acquire the resource (but if the semaphore limit is reached, the executor thread will be blocked and de-prioritized in the thread queue and donating its priority of the child processes).
- The task creator needs it so that it can give it to the task to release the resource when the task is complete. The executor can't check in on the child process since the OS effectively blocks the executor thread knowing that it needs the child processes to complete due to the reason mentioned above.

Reply:
^ That's the strategy that was adopted in the backend tests.

Thread creation: https://github.com/oppia/oppia/blob/develop/scripts/run_backend_tests.py#L184

Thread execution: https://github.com/oppia/oppia/blob/develop/scripts/run_backend_tests.py#L255

Reply:
Thanks, @kevinlee12 for clarification.